games/dots/action.py

```diff
@@ -0,0 +1,79 @@
+import itertools
+from copy import deepcopy
+from numpy.typing import NDArray
+from typing import List, Tuple
+from dataclasses import dataclass
+
+from base.action import Action
+from games.dots.state import DotsState, Chain
+
+
+@dataclass
+class DotsAction(Action):
+    color: str
+    row: int
+    col: int
+
+    def __hash__(self):
+        return hash((self.color, self.row, self.col))
+
+    def apply(self, state: DotsState) -> DotsState:
+        new_board = deepcopy(state.board)
+        new_board[self.row, self.col].color = self.color
+        new_chains = deepcopy(state.chains)
+        for chain in self._find_new_chains(new_board):
+            if chain not in new_chains:
+                if self._mark_surrounded(new_board, chain):
+                    new_chains.append(chain)
+        return DotsState(board=new_board, chains=new_chains)
+
+    def _neighbours(self, board: NDArray, vertex: Tuple[int, int]) -> List[Tuple[int, int]]:
+        neighbours = []
+        for x in range(vertex[0] - 1, vertex[0] + 2):
```

Comment:
Trochę nie rozumiem z czego wynikają te -1 i +2

Reply:
Sąsiadem (x, y) mogą być wierzchołki w jego najbliższym otoczeniu czyli o współrzędnych (x-1, y-1), (x, y-1), (x+1, y-1), (x-1, y) itd

Reply:
Faktycznie. To +2 wprowadziło mnie w błąd mylące, ale już rozumiem z czego wynika 😅